Java/core/src/main/java/com/amazon/randomcutforest/imputation/Summarizer.java

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.imputation;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static java.lang.Math.min;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.function.BiFunction;
+
+import com.amazon.randomcutforest.returntypes.SampleSummary;
+
+public class Summarizer {
+
+    /**
+     * the following determines the ratio between the sum of the (average) radius
+     * and the sepration between centers for a merge; ratio greater than 1 means
+     * significant overlap a ratio of 0 means merge closest pairs without
+     * consideration of separartion
+     *
+     **/
+    public static double SEPARATION_RATIO_FOR_MERGE = 0.8;
+
+    public static int PHASE2_THRESHOLD = 2;
+
+    public static int LENGTH_BOUND = 1000;
+
+    /**
+     * adds a center corresponding to point number index in the list, provided there
+     * is already no other center with the same coordinates (or rather, ruling out
+     * distance 0 from other centers)
+     * 
+     * @param points  a list of points
+     * @param centers a list of current centers
+     * @param dist    distance function
+     * @param index   the index of the point to be added as a center
+     */
+    public static void addCenter(ArrayList<ProjectedPoint> points, ArrayList<Center> centers,
+            BiFunction<float[], float[], Double> dist, int index) {
+        double minDist = Double.MAX_VALUE;
+        for (int i = 0; i < centers.size(); i++) {
+            double t = dist.apply(centers.get(i).coordinate, points.get(index).coordinate);
+            minDist = min(t, minDist);
+        }
+        if (minDist > 0.0) {
+            centers.add(new Center(points.get(index).coordinate, points.get(index).weight));
+        }
+    }
+
+    /**
+     * picks an item such that the prefix sum of the weight of the projectedPoints
+     * is at most wt
+     * 
+     * @param points a list of projected points (points with weight)
+     * @param wt     a selection parameter
+     * @return the index of the point for which the condition is satisfied (or the
+     *         last index)
+     */
+    public static int pick(ArrayList<ProjectedPoint> points, double wt) {
```

Comment:
This is a very generic method name for very specialized functionality.

Reply:
fixed.